File: weathersdk/src/main/java/fr/ekito/myweatherlibrary/WeatherService.java
package fr.ekito.myweatherlibrary;

import android.app.Service;
import android.content.Intent;
import android.os.Binder;
import android.os.IBinder;

import fr.ekito.myweatherlibrary.di.Injector;
import fr.ekito.myweatherlibrary.json.geocode.Geocode;
import fr.ekito.myweatherlibrary.json.weather.Weather;
import fr.ekito.myweatherlibrary.ws.WeatherWS;
import rx.Observable;

public class WeatherService extends Service {

    private static final String TAG = WeatherService.class.getSimpleName();
    private final IBinder mBinder = new LocalBinder();
    private WeatherWS weatherWS;

    public WeatherService() {
    }

    public Observable<Geocode> geocode(String address) {
        return weatherWS.geocode(address);
    }

    public Observable<Weather> weather(Double lat, Double lon, String lang) {
        return weatherWS.weather(lat, lon, lang);
    }

    @Override
    public IBinder onBind(Intent intent) {
        // inject stuff
        weatherWS = Injector.get(WeatherWS.class);

        return mBinder;
    }

    @Override
    public boolean onUnbind(Intent intent) {
        // close stuff
        return super.onUnbind(intent);
    }

    public class LocalBinder extends Binder {
        public WeatherService getService() {
            return WeatherService.this;
        }
    }
}

File: weathersdk/src/main/kotlin/fr/ekito/myweatherlibrary/WeatherService.kt
package fr.ekito.myweatherlibrary

import android.app.Service
import android.content.Intent
import android.os.Binder
import android.os.IBinder
import fr.ekito.myweatherlibrary.di.KInjector
import fr.ekito.myweatherlibrary.json.geocode.Geocode
import fr.ekito.myweatherlibrary.json.weather.Weather
import fr.ekito.myweatherlibrary.ws.WeatherWS
import rx.Observable

/**
 * Weather service
 */
class WeatherService : Service() {

    val mBinder = LocalBinder()
    lateinit var weatherWS: WeatherWS

    override fun onBind(intent: Intent): IBinder? {
        // inject stuff
        weatherWS = KInjector.weatherWS
        return mBinder
    }

    override fun onUnbind(intent: Intent): Boolean {
        // close stuff
        return super.onUnbind(intent)
    }

    inner class LocalBinder : Binder() {
        val service: WeatherService
            get() = this@WeatherService
    }

    /**
     * Get geocode for given address
     */
    fun geocode(address: String): Observable<Geocode> {
        return weatherWS.geocode(address)
    }

    /**
     * Get weather for given gps location
     * @see geocode to get gps location for given address
     */
    fun weather(lat: Double, lon: Double, lang: String): Observable<Weather> {
        return weatherWS.weather(lat, lon, lang)
    }
}
